File: src/main/java/me/wawwior/utils/hashing/Salt.java
package me.wawwior.utils.hashing;

import java.util.Random;

public interface Salt {

    static Salt random(int length) {
        return random(length, new Random());
    }

    static Salt random(int length, long seed) {
        return random(length, new Random(seed));
    }

    static Salt random(int length, Random random) {
        return new Salt() {
            @Override
            public byte[] getSalt() {
                byte[] salt = new byte[length];
                random.nextBytes(salt);
                return salt;
            }

            @Override
            public int getLength() {
                return length;
            }
        };
    }

    static Salt fixed(byte[] salt) {
        return new Salt() {
            @Override
            public byte[] getSalt() {
                return salt;
            }

            @Override
            public int getLength() {
                return salt.length;
            }
        };
    }

    static Salt none() {
        return fixed(new byte[0]);
    }

    byte[] getSalt();

    int getLength();

}

File: src/main/kotlin/me/wawwior/utils/hashing/Salt.kt
package me.wawwior.utils.hashing

import java.util.*

interface Salt {

    companion object {

        @JvmStatic
        fun random(length: Int): Salt {
            return random(length, Random())
        }

        @JvmStatic
        fun random(length: Int, seed: Long): Salt {
            return random(length, Random(seed))
        }

        private fun random(length: Int, random: Random): Salt {
            return object : Salt {
                override fun getSalt(): ByteArray {
                    val salt = ByteArray(length)
                    random.nextBytes(salt)
                    return salt
                }

                override fun getLength(): Int {
                    return length
                }
            }
        }

        @JvmStatic
        fun fixed(salt: ByteArray): Salt {
            return object : Salt {
                override fun getSalt(): ByteArray {
                    return salt
                }

                override fun getLength(): Int {
                    return salt.size
                }
            }
        }

        @JvmStatic
        fun none(): Salt {
            return fixed(ByteArray(0))
        }
    }

    fun getSalt(): ByteArray

    fun getLength(): Int

}